eth/stagedsync/stage_cumulative_index.go

```diff
@@ -0,0 +1,147 @@
+package stagedsync
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon/common"
+	"github.com/ledgerwatch/erigon/common/dbutils"
+	"github.com/ledgerwatch/erigon/core/rawdb"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/eth/stagedsync/stages"
+	"github.com/ledgerwatch/erigon/rlp"
+	"github.com/ledgerwatch/log/v3"
+)
+
+type CumulativeIndexCfg struct {
+	db kv.RwDB
+}
+
+func StageCumulativeIndexCfg(db kv.RwDB) CumulativeIndexCfg {
+	return CumulativeIndexCfg{
+		db: db,
+	}
+}
+
+func SpawnStageCumulativeIndex(cfg CumulativeIndexCfg, s *StageState, tx kv.RwTx, ctx context.Context) error {
+	useExternalTx := tx != nil
+
+	if !useExternalTx {
+		var err error
+		tx, err = cfg.db.BeginRw(context.Background())
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback()
+	}
+	// Log timer
+	logEvery := time.NewTicker(logInterval)
+	defer logEvery.Stop()
+
+	headNumber, err := stages.GetStageProgress(tx, stages.Headers)
+	if err != nil {
+		return fmt.Errorf("getting bodies progress: %w", err)
+	}
+	// If we are done already, we can exit the stage
+	if s.BlockNumber == headNumber {
+		return nil
+	}
+
+	cumulativeGasUsed, err := rawdb.ReadCumulativeGasUsed(tx, s.BlockNumber)
+	if err != nil {
+		return err
+	}
+
+	prevProgress := s.BlockNumber
+	headerC, err := tx.Cursor(kv.Headers)
+	if err != nil {
+		return err
+	}
+
+	currentBlockNumber := s.BlockNumber + 1
+	for k, v, err := headerC.Seek(dbutils.EncodeBlockNumber(s.BlockNumber)); k != nil; k, v, err = headerC.Next() {
+		if err != nil {
+			return err
+		}
+
+		if len(k) != 40 {
+			continue
+		}
+
+		blockNumber, err := dbutils.DecodeBlockNumber(k[:8])
+		if err != nil {
+			return err
+		}
+
+		canonicalHash, err := rawdb.ReadCanonicalHash(tx, blockNumber)
+		if err != nil {
+			return err
+		}
+
+		if canonicalHash != common.BytesToHash(k[8:]) {
+			continue
+		}
+
+		currentBlockNumber = blockNumber
+
+		var header types.Header
+		if err := rlp.Decode(bytes.NewReader(v), &header); err != nil {
+			log.Error("Invalid block header RLP", "number", currentBlockNumber, "err", err)
+			return nil
```

Comment:
why not return err here? then it will be also logged by log.Error but also will have log prefix of stage.